DA-140: add mouse mover tutorial (#927)
viamrobotics/docs: docs/tutorials/get-started/single-component-tutorials-servo-mousemover.md

</div>

This tutorial will show you how to build a mouse mover using Viam, a Raspberry Pi, a servo, and an optical mouse.
This robot will turn the continuous servo that's secured inside the box, which will turn the circle under the optical mouse. This will keep your computer from falling asleep.

This project is a good place to begin if you're new to robots and would like to learn how to use a [servo component](/components/servo/) with the Viam app and Viam's [Python SDK](https://python.viam.dev/).

<div class="td-max-width-on-larger-screens">
  {{<gif webm_src="../../img/single-component-tutorials-servo-mousemover/moving.webm" mp4_src="../../img/single-component-tutorials-servo-mousemover/moving.mp4" alt="Video of a mouse running erratically on the screen in a sweeping motion in front of the Viam app Control page with an overlaid video of a mouse on top of a cardboard box with a moving cardboard circle underneath it with red swirl lines on the circle.">}}
</div>

> {{< alert title="Tip" color="tip" >}}
> NOTE: This tutorial is part of Viam's Single Component Tutorial series, where you have the opportunity to learn how to use single components for low cost with Viam's platform in fun mini-tutorials.
> 
> Stay tuned for more tutorials like this one and don't forget to [join our Discord](https://discord.gg/viam) if you have any questions in the meantime!
> {{< /alert >}}

Reviewer: Please remove this for now. This isn't a series yet - and it's confusing for people to see this touted as a series. I understand it is meant to become one. Once it is you can add this here but for now it shouldn't be here. 
```suggestion
```

- microSD card reader
- [Continuous rotation servo](https://a.co/d/2w0u6rK) (we used the FS90R servo)
- Wheel or arm for servo (this comes in some of the FS90R packages)
- [Jumper wires](https://www.amazon.com/EDGELEC-Breadboard-Optional-Assorted-Multicolored/dp/B07GD2BWPY)
- Optical mouse - corded or uncorded

### Tools and Consumables

- Suitable sized box (box should allow servo with wheel/arm with circle cut out on top of wheel/arm to sit flush with the top of the box)
- Double sided tape
- Tape to seal the box (we used black gaffers tape)
- Box cutters
- Marker

### Software

- [Python3](https://www.python.org/download/releases/3.0/)
- [Pip](https://pip.pypa.io/en/stable/#)
- [viam-server](https://github.com/viamrobotics/rdk/tree/0c550c246739b87b4d5a9e8d96d2b6fdb3948e2b)
- [Viam Python SDK](https://python.viam.dev/)
- [Mousemover tutorial code](https://github.com/viam-labs/tutorial-mousemover)

## Install Viam software

First, install viam-server according to our [Raspberry Pi setup guide](/installation/prepare/rpi-setup/)

Next, run this command in your Raspberry Pi terminal to install the pip package manager. Select "yes" when it asks if you want to continue.

> `sudo apt-get install pip`

Reviewer: ```suggestion
```sh
sudo apt-get install pip
```
```

The command above installs the latest version of `python3` and `pip3` on your Raspberry Pi.
To verify and get the version of the package, you can run the command:

> `pip3 --version`

Reviewer: ```suggestion
```sh
pip3 --version
```
```

The [Viam Python SDK](https://python.viam.dev/) (Software Development Kit) allows you to write programs in the Python programming language to operate robots using Viam. To get the Python SDK working on the Raspberry Pi, run the following command in your Raspberry Pi terminal:

> `pip install viam-sdk`

Reviewer: ```suggestion
```sh
pip install viam-sdk
```
```

![A screenshot of the word Language with four boxes below it with Python, Golang, Typescript (Web), and Remotes in each one. Python has a black background with white text and a checkmark, the other three have white backgrounds with black text.](../../img/single-component-tutorials-servo-mousemover/choose-python.png)

Since we already installed the Python SDK, we can skip the first step. Copy the code in from step 2.

![A screenshot of the Viam app showing the CODE SAMPLE tab. On this tab it has Language: with four boxes, Python is black background with white text and has a checkmark indicating it is selected, then there's Golang, Typescript (Web), and Remotes with white backgrounds and black text. The page next says Python SDK (Documentation) with a link to Documentation. Below that there\'s a place that says "pip install viam-SDK" this is how you install the Viam Python SDK. Below that instruction is the instructions to "Then copy and run the following code to connect to your robot:" with 7 lines of code seen (the screenshot is incomplete).](../../img/single-component-tutorials-servo-mousemover/code-sample-page.png)

> The copied the code needs to go in a Python file on the Raspberry Pi.

Reviewer: ```suggestion
The copied code needs to go in a Python file on the Raspberry Pi.
```

You can do so by creating a file directly inside the Raspberry Pi and editing the file with nano.

Inside the Raspberry Pi Terminal, run the following command to create a folder to put our files in (name this folder whatever you want).

`mkdir mousefolder`

Go into mousefolder.

> `cd mousefolder`

Reviewer: ```suggestion
```sh
cd mousefolder
```
```

Create a file using nano with the .py which is the python file extension, (name this file whatever you want).

> `nano anyname.py`

Reviewer: ```suggestion
```sh
nano anyname.py
```
```

Paste the code you got from the **CODE SAMPLE** tab in the Viam app. Press CTRL+O, then CTRL+M, then CTRL+X to save the code and exit.

> ![Raspberry pi computer terminal with black background and mostly white, but there's some green, teal, purple and red parts of the text. At the top it says GNU nano 5.4 anyname.py. Inside the terminal it shows the connect code from the previous example, but this time it's in the terminal. At the bottom it shows the settings for this nano program.](../../img/single-component-tutorials-servo-mousemover/sample-code.png)

Reviewer: ```suggestion
![Raspberry pi terminal showing the connect code from the previous example.](../../img/single-component-tutorials-servo-mousemover/sample-code.png)
```

(Alt texts should have the important infor the picture is trying to convey)

Now, run the code using the below command to get the resource information that lets us know if the connection is good or if there are any errors.

> `python3 anyname.py`

Reviewer: ```suggestion
```sh
python3 anyname.py
```
```

> ![Raspberry pi computer terminal with black background and mostly white, but there's one teal word: "INFO" located at the top next to viam.rpc.dial (dial.py:209). The code shown below that is white text and shows Viam's RDK Resources, and it shows no error.](../../img/single-component-tutorials-servo-mousemover/resources-printed.png)

Reviewer: ```suggestion
![The output of the code showing a list of the resources.](../../img/single-component-tutorials-servo-mousemover/resources-printed.png)
```

> There are no errors in the resources above, so we have confirmed a good software connection between the Python SDK and your Viam robot!

Reviewer: ```suggestion
If your program outputs no errors, your code can connect to your robot correctly.
```

This prints out resources that the user hasn't configured yet. Also I cannot imagine what an error in the resource would be here. The suggestion makes that clearer

## Connect the Servo

First, **turn off your Raspberry Pi**.
This will help prevent any accidents when connecting the hardware.

Now, follow this schematic to attach jumper wires to the servo and Raspberry Pi.

- **GND** on servo attached to pin 6 on Pi
- **5V** on servo attached to pin 2 on Pi
- **PWM** on servo attached to pin 12 on Pi

> ![Computer drawn photo schematic of a raspberry pi connected to a FS90R servo. It shows the yellow wire coming from the servo as the PWM wire and it's attached to pin twelve on the raspberry pi. The red wire coming from the servo is five volt wire and it's attached to pin two. The black wire coming from the servo is the ground wire and it's attached to pin eight.](../../img/single-component-tutorials-servo-mousemover/servo-wiring-diagram.png)

Reviewer: ```suggestion
![A raspberry pi connected to a FS90R servo. It shows the yellow wire coming from the servo as the P W M wire and it's attached to pin twelve on the raspberry pi. The red wire coming from the servo is five volt wire and it's attached to pin two. The black wire coming from the servo is the ground wire and it's attached to pin eight.](../../img/single-component-tutorials-servo-mousemover/servo-wiring-diagram.png)
```

^ there are more issues with this alt text but let's leave it at that

Once you have the wires connected, attach the wheel/arm to servo and turn on the Raspberry Pi.

## Configure your robot

The servo is now physically connected to the Raspberry Pi, but the Viam app hasn't been told the details of which components/services it's using yet, so it's not able to control the servo.

Go to the [Viam app](https://app.viam.com), and navigate to the **CONFIG** tab.

> ![This is a screenshot of the Viam app on the CONFIG page. It shows that Builder is chosen as the mode. Below that in the components tab it shows Create Component and to the far left It says Name with a box for a Name to be typed in. The next box to the right says Type and it shows a drop down list of different types of components in Viam. To the far right it says Model, but it's greyed out because none of the previous two fields have been filled in. Directly to the right of Model is a Create Component button that will create the component.](../../img/single-component-tutorials-servo-mousemover/new-component.png)

Reviewer: ```suggestion
![The Viam app config page on the components tab.](../../img/single-component-tutorials-servo-mousemover/new-component.png)
```

</div>

### Control your robot with code

Copy the code from the [mousemover Github repository](https://github.com/viam-labs/tutorial-mousemover) into your nano file, save it, and run it.

The code uses the Python SDK to securely connect to your robot through Viam app.
Then, it enters a for loop in which **position** tells us the servo **myServo** to move to positions (angle, or speed/direction) between 80 and 93 degrees as specified in the **sequence** list.
The code uses **pause_time** to wait for a random amount of time between 5 and 20 seconds to stay at that position.

You can adjust the range (or speed/direction) by changing the two numbers in the position statement. (Currently set at 80 and 93).
You can also change the time that the servo is allowed to spin in any direction by changing the two numbers in the pause_time statement.
Experiment and have fun.

## Next steps

If you want more, head over to our documentation site and take a look at our [tutorials](/tutorials/).

> If you liked this tutorial and want to do more tutorials like these, then keep checking back for more single component tutorials.

Reviewer: ```suggestion
```
